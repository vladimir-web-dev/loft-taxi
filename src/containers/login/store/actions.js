import { createAction } from 'redux-actions';

export const authRequest = createAction('taxi/auth/AUTH_REQUEST');
export const authSuccess = createAction('taxi/auth/AUTH_SUCCESS');
export const authFailure = createAction('taxi/auth/AUTH_FAILURE');

export const registrationRequest = createAction('taxi/auth/REGISTRATION_REQUEST');
export const registrationSuccess = createAction('taxi/auth/REGISTRATION_SUCCESS');
export const registrationFailure = createAction('taxi/auth/REGISTRATION_FAILURE');

export const logout = createAction('taxi/auth/LOGOUT');

