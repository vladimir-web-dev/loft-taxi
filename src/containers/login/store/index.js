export { default } from './reducer';
export * from './actions';
export * from './selectors';
export * from './middlewares';
export * from './sagas';