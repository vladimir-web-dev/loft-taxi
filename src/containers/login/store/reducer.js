import {
    authRequest,
    authSuccess,
    authFailure,
    registrationRequest,
    registrationSuccess,
    registrationFailure,
    logout
} from './actions';
import { handleActions } from 'redux-actions';
import { combineReducers } from 'redux';

const token = handleActions({
    [authRequest]: () => "",
    [authSuccess]: (_state, action) => action.payload,
    [registrationRequest]: () => "",
    [registrationSuccess]: (_state, action) => action.payload,
    [logout]: () => ""
}, "");

const isAuthenticating = handleActions({
    [authRequest]: () => true,
    [authSuccess]: () => false,
    [authFailure]: () => false,
    [registrationRequest]: () => true,
    [registrationSuccess]: () => false,
    [registrationFailure]: () => false
}, false);

const isAuthenticated = handleActions({   
    [authSuccess]: () => true,
    [authFailure]: () => false,  
    [registrationSuccess]: () => true,
    [registrationFailure]: () => false,
    [logout]: () => false
}, false);

const authError = handleActions({   
    [authSuccess]: () => "",
    [authFailure]: (_state, action) => action.payload,  
}, "");

const registrationError = handleActions({    
    [registrationSuccess]: () => "",
    [registrationFailure]: (_state, action) => action.payload
}, "");

export default combineReducers({isAuthenticating, isAuthenticated, token, authError, registrationError});
