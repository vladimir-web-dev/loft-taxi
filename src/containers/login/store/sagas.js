import {
  authRequest,
  authSuccess,
  authFailure,
  registrationRequest,
  registrationSuccess,
  registrationFailure,
  logout
} from "./actions";
import { apiCallsHelper } from "../../../helpers";
import { take, call, put, takeLatest, all, fork } from "redux-saga/effects";

const URL_LOGIN = "https://loft-taxi.glitch.me/auth";
const URL_REGISTER = "https://loft-taxi.glitch.me/register";

function* fetchLoginData({ data }) {
  return yield call(apiCallsHelper.post, URL_LOGIN, data);
}

function* fetchRegistrationData({ data }) {
  yield call(apiCallsHelper.post, URL_REGISTER, data);
}

function* watchRegistrationRequest() {
  const { success, token } = yield takeLatest(
    registrationRequest.toString(),
    fetchRegistrationData
  );
  if (success) {
    put(registrationSuccess(token));
  } else {
    put(registrationFailure(token));
  }
}

function* watchLoginLogoutRequest() {
  while (true) {
    const { payload } = yield take(authRequest.toString());
    const { success, token } = yield call(fetchLoginData, payload);

    if (success) {
      yield put(authSuccess(token));
      payload.history.push("/map");
      yield take(logout.toString());
    } else {
      yield put(authFailure(token));
    }
  }
}

export function* authSaga() {
  yield all([
    yield fork(watchRegistrationRequest),
    yield fork(watchLoginLogoutRequest)
  ]);
}
