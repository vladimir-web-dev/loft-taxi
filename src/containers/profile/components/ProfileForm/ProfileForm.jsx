import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import TextField from "@material-ui/core/TextField";
import { DatePicker, MuiPickersUtilsProvider } from "@material-ui/pickers";
import MomentUtils from "@date-io/moment";
import Button from "@material-ui/core/Button";
import {
  getCardSelector,
  getTokenSelector,
  cardUpdateRequest,
  cardDetailsRequest
} from "../../../profile/store";
import { useForm, Controller } from "react-hook-form";
import formStyles from "../../../../css_modules/Form.module.css";
import classNames from 'classnames';

export function ProfileForm() {
  const { setValue, handleSubmit, control, errors, setError, clearError } = useForm();
  const hasCardNumberError = errors.cardNumber !== undefined;
  const cardNumberError = errors.cardNumber && errors.cardNumber.message;
  const hasCVCNumberError = errors.cvc !== undefined;
  const cvcNumberError = errors.cvc && errors.cvc.message;
  const dispatch = useDispatch();
  const token = useSelector(state => getTokenSelector(state.authReducer));
  const card = useSelector(state => getCardSelector(state.profileReducer));
  

  if (Object.keys(card).length > 0) {
    setValue("expiryDate", card.expiryDate);
    setValue("cardName", card.cardName);
    setValue("cvc", card.cvc);
    setValue("cardNumber", card.cardNumber);
  }

  const onSubmit = data => {
    if(!validateCardNumber(data.cardNumber) || !validateCVCNumber(data.cvc))
      return;

    const payload = {
      ...data,
      token
    };

    dispatch(cardUpdateRequest(payload));
  };

  const validateCVCNumber = value => {    

    if(value.length !== 3) {
      setError("cvc","notMatch","В cvc номере должно быть 3 цифры");
      return false;
    }   

    clearError("cvc");
    return true;
      
  }

  const validateCardNumber = value => {    
    const reg = new RegExp('^[0-9]+$');

    if(value.length !== 16) {
      setError("cardNumber","notMatch","В номере карты 16 цифр");
      return false;
    }   
    
    if(!reg.test(value)) {
      setError("cardNumber","notMatch","В номере карты должны быть только цифры");
      return false;
    }

    clearError("cardNumber");
    return true;
      
  }

  useEffect(() => {
    dispatch(cardDetailsRequest(token));
  }, []);

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className={classNames([formStyles.container, formStyles.formCentered])}>
        <h1 className={formStyles.heading}>Профиль</h1>
        <span className={formStyles.info}>Способ оплаты</span>
        <div className={formStyles.formRow}>
          <div className={formStyles.formCol}>
            <div className={formStyles.card}>
              <div className={formStyles.formRow}>
                <Controller                 
                  as={TextField}
                  control={control}
                  data-testid="input-card-number"
                  name="cardNumber"
                  label="Номер карты"
                  placeholder="0000 0000 0000 0000"
                  error={hasCardNumberError}
                  helperText={cardNumberError}
                  defaultValue=""
                  margin="dense"
                  required
                  fullWidth
                />
              </div>
              <div className={formStyles.formRow}>
                <MuiPickersUtilsProvider utils={MomentUtils} >
                  <Controller
                  style={{marginTop: '20px', marginBottom:'4px'}}
                  fullWidth
                    as={DatePicker}
                    name="expiryDate"
                    control={control}
                  />
                </MuiPickersUtilsProvider>
              </div>
            </div>
          </div>
          <div className={formStyles.formCol}>
            <div className={formStyles.card}>
              <div className={formStyles.formRow}>
                <Controller
                  as={TextField}
                  control={control}
                  data-testid="input-card-number"
                  name="cardName"
                  label="Имя владельца"
                  placeholder="USER NAME"
                  defaultValue=""
                  margin="dense"
                  required
                  fullWidth
                />
              </div>
              <div className={formStyles.formRow}>
                <Controller
                  as={TextField}
                  control={control}
                  data-testid="input-card-number"
                  name="cvc"
                  label="CVC "
                  placeholder="CVC"
                  error={hasCVCNumberError}
                  helperText={cvcNumberError}
                  defaultValue=""
                  margin="dense"
                  required
                  fullWidth
                />
              </div>
            </div>
          </div>
        </div>
        <Button
          data-testid="button-login"
          color="secondary"
          variant="contained"
          type="submit"
        >
          Сохранить
        </Button>
      </div>
    </form>
  );
}
