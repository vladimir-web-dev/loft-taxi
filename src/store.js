import { createStore, applyMiddleware, compose } from 'redux';
import authReducer, { mwAuthentication, mwRegistration } from './modules/auth';
import cardReducer, { mwCard } from './modules/card';
import { combineReducers } from 'redux';


export const initStore = () => { 
    return createStore(
        combineReducers({
            authReducer, 
            cardReducer
        }),
        compose(
            applyMiddleware(mwAuthentication),
            applyMiddleware(mwRegistration),
            applyMiddleware(mwCard)
        )
    );
};
    