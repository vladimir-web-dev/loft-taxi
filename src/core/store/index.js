import { createStore, applyMiddleware, compose } from "redux";
import { rootReducer } from "../reducer";
import { rootSaga } from "../saga";
import createSagaMiddleware from "redux-saga";

export function initStore() {
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore(rootReducer, compose(
    applyMiddleware(sagaMiddleware)
    ));

  sagaMiddleware.run(rootSaga);

  return store;
}
