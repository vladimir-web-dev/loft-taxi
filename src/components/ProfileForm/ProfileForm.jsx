import React, { useEffect } from "react";
import Paper from "@material-ui/core/Paper";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import { useTheme } from "@material-ui/core/styles";
import { cardUpdateRequest, cardDetailsRequest } from "../../modules/card";
import { useDispatch, useSelector } from "react-redux";
import TextField from "@material-ui/core/TextField";
import { DatePicker, MuiPickersUtilsProvider } from "@material-ui/pickers";
import MomentUtils from "@date-io/moment";
import Button from "@material-ui/core/Button";
import { getCard, getToken, getIsUpdated } from "../../modules/card";
import { useForm, Controller } from "react-hook-form";
import { Link } from "react-router-dom";

export function ProfileForm() {
  const { register, setValue, handleSubmit, control } = useForm();

  const theme = useTheme();
  const dispatch = useDispatch();
  const token = useSelector(state => getToken(state.authReducer));
  const card = useSelector(state => getCard(state.cardReducer));
  const isUpdated = useSelector(state => getIsUpdated(state.cardReducer));

  if (Object.keys(card).length > 0) {
    setValue("expiryDate", card.expiryDate);
    setValue("cardName", card.cardName);
    setValue("cvc", card.cvc);
    setValue("cardNumber", card.cardNumber);
  }

  const onSubmit = (data, e) => {
    e.preventDefault();
    const payload = {
      ...data,
      token
    };
    
    dispatch(cardUpdateRequest(payload));
  };

  useEffect(() => {
    
    dispatch(cardDetailsRequest(token));
  }, []);

  return (
    <Paper elevatio={1} style={{ padding: theme.spacing(6, 6), marginTop: theme.spacing(6) }}>
      <Grid container alignItems="center" direction="column">
        <Grid item>
          <Typography variant="h3" component="h1" gutterBottom>
            Профиль
          </Typography>
        </Grid>
        {isUpdated ? (
          <>
            <Grid item style={{ marginBottom: theme.spacing(4) }}>
              <Typography variant="h5" component="h5" gutterBottom>
                Платёжные данные обновлены. Теперь вы можете заказывать такси.
              </Typography>
            </Grid>
            <Grid item xs={12} container justify="center">
              <Button 
                component={Link} to='/map' 
                color="secondary"
                variant="contained">Перейти на карту</Button>
            </Grid>
          </>
        ) : (
          <form
            data-testid="login-form"
            className="formTag"
            onSubmit={handleSubmit(onSubmit)}
          >
            <Grid item container xs={12} style={{ marginBottom: theme.spacing(4) }} spacing={3}>
              <Grid item container xs={6}>
                <Paper elevatio={1} style={{ padding: theme.spacing(2, 2) }}>
                  <Grid container direction="column">
                    <Grid item xs={12}>
                      <TextField
                        inputRef={register}
                        data-testid="input-card-number"
                        name="cardNumber"
                        label="Номер карты"
                        placeholder="0000 0000 0000 0000"
                        margin="dense"
                        required
                        fullWidth
                      />
                    </Grid>
                    <Grid item xs={12}>
                      <MuiPickersUtilsProvider utils={MomentUtils}>
                        <Controller
                          as={DatePicker}
                          name="expiryDate"
                          control={control}
                        />
                      </MuiPickersUtilsProvider>
                    </Grid>
                  </Grid>
                </Paper>
              </Grid>
              <Grid item container xs={6}>
                <Paper elevatio={1} style={{ padding: theme.spacing(2, 2) }}>
                  <Grid container direction="column">
                    <Grid item xs={12}>
                      <TextField
                        inputRef={register}
                        data-testid="input-card-number"
                        name="cardName"
                        label="Имя владельца"
                        placeholder="USER NAME"
                        margin="dense"
                        required
                        fullWidth
                      />
                    </Grid>
                    <Grid item xs={12}>
                      <TextField
                        inputRef={register}
                        data-testid="input-card-number"
                        name="cvc"
                        label="CVC "
                        placeholder="CVC "
                        margin="dense"
                        required
                        fullWidth
                      />
                    </Grid>
                  </Grid>
                </Paper>
              </Grid>
            </Grid>
            <Grid item xs={12} container justify="center">
              <Button
                data-testid="button-login"
                color="secondary"
                variant="contained"
                type="submit"
              >
                Сохранить
              </Button>
            </Grid>
          </form>
        )}
      </Grid>
    </Paper>
  );
}
