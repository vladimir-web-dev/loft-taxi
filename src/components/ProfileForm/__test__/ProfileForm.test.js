import React from 'react';
import ReactDOM from 'react-dom';
import {ProfileForm} from '../ProfileForm';
import {render} from '@testing-library/react';
import { Provider } from 'react-redux'
import { initStore } from "../../../store";
import 'mutationobserver-shim';
global.MutationObserver = window.MutationObserver;


it('renders without crashing', () => {
    const store = initStore();
    const div = document.createElement('div');

    ReactDOM.render(
        <Provider store={store}>
            <ProfileForm />
        </Provider>, div);
    ReactDOM.unmountComponentAtNode(div);
})

it('matches snapshot', () => {
    const store = initStore();
    const div = document.createElement('div');

    const {container} = render(
    <Provider store={store}>
        <ProfileForm />
    </Provider>, div);
    expect(container).toMatchSnapshot(); 
})