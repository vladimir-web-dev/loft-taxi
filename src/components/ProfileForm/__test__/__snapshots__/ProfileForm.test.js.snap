// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    elevatio="1"
    style="padding: 48px 48px; margin-top: 48px;"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <h1
          class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom"
        >
          Профиль
        </h1>
      </div>
      <form
        class="formTag"
        data-testid="login-form"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-item MuiGrid-grid-xs-12"
          style="margin-bottom: 32px;"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-6"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
              elevatio="1"
              style="padding: 16px 16px;"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
                >
                  <div
                    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
                    data-testid="input-card-number"
                  >
                    <label
                      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense Mui-required Mui-required"
                      data-shrink="false"
                    >
                      Номер карты
                      <span
                        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                      >
                         
                        *
                      </span>
                    </label>
                    <div
                      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
                        name="cardNumber"
                        placeholder="0000 0000 0000 0000"
                        required=""
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
                >
                  <div
                    class="MuiFormControl-root MuiTextField-root"
                    format="MMMM Do"
                  >
                    <div
                      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiInputBase-input MuiInput-input"
                        name="expiryDate"
                        readonly=""
                        type="text"
                        value="February 7th"
                      />
                    </div>
                    
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-6"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
              elevatio="1"
              style="padding: 16px 16px;"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
                >
                  <div
                    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
                    data-testid="input-card-number"
                  >
                    <label
                      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense Mui-required Mui-required"
                      data-shrink="false"
                    >
                      Имя владельца
                      <span
                        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                      >
                         
                        *
                      </span>
                    </label>
                    <div
                      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
                        name="cardName"
                        placeholder="USER NAME"
                        required=""
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
                >
                  <div
                    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
                    data-testid="input-card-number"
                  >
                    <label
                      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense Mui-required Mui-required"
                      data-shrink="false"
                    >
                      CVC 
                      <span
                        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                      >
                         
                        *
                      </span>
                    </label>
                    <div
                      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
                        name="cvc"
                        placeholder="CVC "
                        required=""
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-center MuiGrid-grid-xs-12"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
            data-testid="button-login"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Сохранить
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
